dcap-ql/src/quote.rs

```diff
@@ -0,0 +1,382 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::borrow::Cow;
+use std::mem;
+
+use byteorder::{ByteOrder, LE};
+use num_traits::FromPrimitive;
+
+// ====================================================
+// ================= TYPE DEFINITIONS =================
+// ====================================================
+
+pub struct Quote<'a> {
+    header: QuoteHeader<'a>,
+    report_body: Cow<'a, [u8]>,
+    signature: Cow<'a, [u8]>,
+}
+
+pub enum QuoteHeader<'a> {
+    V3 {
+        attestation_key_type: Quote3AttestationKeyType,
+        qe3_svn: u16,
+        pce_svn: u16,
+        qe3_vendor_id: Cow<'a, [u8]>,
+        user_data: Cow<'a, [u8]>,
+    },
+}
+
+#[repr(u16)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, FromPrimitive, ToPrimitive)]
+pub enum Quote3AttestationKeyType {
+    EcdsaP256 = 2,
+    EcdsaP384 = 3,
+}
+
+pub const QE3_VENDOR_ID_INTEL: [u8; 16] = [
+    0x93, 0x9a, 0x72, 0x33, 0xf7, 0x9c, 0x4c, 0xa9, 0x94, 0x0a, 0x0d, 0xb3, 0x95, 0x7f, 0x06, 0x07,
+];
+
+pub type QeId<'a> = Cow<'a, [u8]>;
+
+pub struct Quote3SignatureEcdsaP256<'a> {
+    signature: Cow<'a, [u8]>,
+    attestation_public_key: Cow<'a, [u8]>,
+    qe3_report: Cow<'a, [u8]>,
+    qe3_signature: Cow<'a, [u8]>,
+    authentication_data: Cow<'a, [u8]>,
+    certification_data_type: CertificationDataType,
+    certification_data: Cow<'a, [u8]>,
+}
+
+#[repr(u16)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, FromPrimitive, ToPrimitive)]
+pub enum CertificationDataType {
+    PpidCleartext = 1,
+    PpidEncryptedRsa2048 = 2,
+    PpidEncryptedRsa3072 = 3,
+    PckCertificate = 4,
+    PckCertificateChain = 5,
+    EcdsaSignatureAuxiliaryData = 6,
+    PlatformManifest = 7,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct Qe3CertDataPpid<'a> {
+    pub ppid: Cow<'a, [u8]>,
+    pub cpusvn: Cow<'a, [u8]>,
+    pub pcesvn: u16,
+    pub pceid: u16,
+}
+
+pub type Result<T> = ::std::result::Result<T, ::failure::Error>;
+
+// ===========================================
+// ================= PARSING =================
+// ===========================================
+
+trait TrySplitAt {
+    fn try_split_at(&self, mid: usize) -> Result<(&Self, &Self)>;
+
+    fn try_split_at_map<'a, F: FnOnce(&'a Self) -> R, R: 'a>(
+        &'a self,
+        mid: usize,
+        f: F,
+    ) -> Result<(R, &'a Self)> {
+        let (a, b) = self.try_split_at(mid)?;
+        Ok((f(a), b))
+    }
+}
+
+impl<T> TrySplitAt for [T] {
+    fn try_split_at(&self, mid: usize) -> Result<(&Self, &Self)> {
+        match (self.get(..mid), self.get(mid..)) {
+            (Some(a), Some(b)) => Ok((a, b)),
+            _ => bail!("Unexpected end of quote"),
+        }
+    }
+}
```

Comment:
Might be easier to use
```
trait TakePrefix: Sized {
    fn take_prefix(&mut self, mid: usize) -> Result<Self>;
}

impl<'a, T: 'a> TakePrefix for &'a [T] {
    fn take_prefix(&mut self, mid: usize) -> Result<Self> {
        if let (Some(prefix), Some(rest)) = (self.get(..mid), self.get(mid..)) {
            *self = rest;
            Ok(prefix)
        } else {
            Err(bail!("Unexpected end of quote"))
        }
    }
}
```

Then e.g. `let (qe3_vendor_id, rest) = rest.try_split_at_map(QE3_VENDOR_ID_LEN, Cow::from)?;` could be `let qe3_vendor_id = quote.take_prefix(QE3_VENDOR_ID_LEN)?.into();`.